README.md
# Aurelia Value Converters

#### Table of Contents
* [Summary](#summary)
* [Download](#download)
* [Sort](#sort)
* [Group](#group)
* [DateFormat](#dateformat)
* [Setup](#setup)

## Summary
A package full of common value converters used across my projects

## Download

`npm install aurelia-value-converters`

main.js
```
import 'bootstrap';

export function configure(aurelia) {
  aurelia.use
    .standardConfiguration()
    .developmentLogging()
    .plugin('aurelia-value-converters');

  aurelia.start().then(() => aurelia.setRoot());
}
```

## Sort
#### Dependencies (lodash.topath)
sort by a property named requested
```
  <!-- Ascending order -->
  <div repeat.for="r of requests | sort:'requested'"></div>

  <!-- Descending order -->
  <div repeat.for="r of requests | sort:'requested':'desc'"></div>

  <!-- sort by a nested property within the bar property -->
  <div repeat.for="r of requests | sort:'bar.requested'"></div>
```

## Group
#### Dependencies (lodash.topath)
group by a property named requested (nested properties supported too)
```
  <div repeat.for="grp of requests | group:'dateGroup'">
    <span class="text-muted">Name of the group: ${grp.group}</span>

    <!-- The group values as an array -->
    <div repeat.for="val of grp.values"></div>

  </div>
```

## DateFormat
#### Dependencies
moment date formats are supported
```
  <span>${date | dateFormat: format || 'MM/DD/YYYY'}</span>
```

## Setup
```
  git clone https://github.com/jmzagorski/aurelia-value-converters.git
  cd aurelia-value-converters
  npm i
  node_modules/.bin/jspm install
  // do work
  node_modules/.bin/jspm test
```


doc/CHANGELOG.md
<a name="1.0.6"></a>
## [1.0.6](https://github.com/jmzagorski/aurelia-value-converters/compare/v1.0.5...v1.0.6) (2018-04-24)


### Bug Fixes

* add lodash.path to handle nested path ([1adc782](https://github.com/jmzagorski/aurelia-value-converters/commit/1adc782))



<a name="1.0.5"></a>
## [1.0.5](https://github.com/jmzagorski/aurelia-value-converters/compare/v1.0.4...v1.0.5) (2017-09-13)



<a name="1.0.4"></a>
## [1.0.4](https://github.com/jmzagorski/aurelia-value-converters/compare/v1.0.3...v1.0.4) (2017-09-13)



<a name="1.0.3"></a>
## [1.0.3](https://github.com/jmzagorski/aurelia-value-converters/compare/v1.0.2...v1.0.3) (2017-05-29)


### Bug Fixes

* **config:** use PLATFORM.moduleName ([e43924b](https://github.com/jmzagorski/aurelia-value-converters/commit/e43924b))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/jmzagorski/aurelia-value-converters/compare/v1.0.1...v1.0.2) (2017-01-26)


### Bug Fixes

* **package:** typo in aurelia resource in package.json ([6ec82f0](https://github.com/jmzagorski/aurelia-value-converters/commit/6ec82f0))





test/setup.js
import 'aurelia-polyfills';

(function() {
  jasmine.setupSpy = function(name, realObj) {
    if (!realObj) throw Error('obj to fake required');
    if (!name) throw Error('name for the fake obj required');

    let keys = [];
    for (let key in realObj) keys.push(key);

    return keys.length > 0 ? jasmine.createSpyObj(name, keys) : {};
  };
})();


test/sort.spec.js
import {SortValueConverter} from '../src/sort';

describe('the sort value converter', () => {
  let sut;

  beforeEach(() => {
    sut = new SortValueConverter();
  });

  [ null,
    undefined,
    []
  ].forEach(array => {
    it('does not throw with no array', () => {
      const ex = () => sut.toView(array);

      expect(ex).not.toThrow();
    });
  });

  [ {propName: null, unsorted: [6, 1, 2], sorted: [1, 2, 6]},
    {propName: undefined, unsorted: [6, 1, 2], sorted: [1, 2, 6]},
    {propName: '', unsorted: [6, 1, 2], sorted: [1, 2, 6]}
  ].forEach(data => {
    it('sorts an array when no property name given', () => {
      const sorted = sut.toView(data.unsorted, data.propName);

      expect(sorted).toEqual(data.sorted);
    });
  });

  [ {dir: 'asc', unsorted: [{id: 3}, {id: 1}, {id: 2} ], sorted: [{id: 1}, {id: 2}, {id: 3}]},
    {dir: '', unsorted: [{id: 3}, {id: 1}, {id: 2} ], sorted: [{id: 1}, {id: 2}, {id: 3}]},
    {dir: null, unsorted: [{id: 3}, {id: 1}, {id: 2} ], sorted: [{id: 1}, {id: 2}, {id: 3}]},
    {dir: undefined, unsorted: [{id: 3}, {id: 1}, {id: 2} ], sorted: [{id: 1}, {id: 2}, {id: 3}]},
    {dir: 'desc', unsorted: [{id: 3}, {id: 1}, {id: 2} ], sorted: [{id: 3}, {id: 2}, {id: 1}]}
  ].forEach(data => {
    it('sorts an number property', () => {
      const sorted = sut.toView(data.unsorted, 'id', data.dir);

      expect(sorted).toEqual(data.sorted);
    });
  });

  // don't need to test different asc dir args since that was tested with the
  // number test above
  [ {dir: 'asc', unsorted: [{id: 'c'}, {id: 'a'}, {id: 'b'} ], sorted: [{id: 'a'}, {id: 'b'}, {id: 'c'}]},
    {dir: 'desc', unsorted: [{id: 'c'}, {id: 'a'}, {id: 'b'} ], sorted: [{id: 'c'}, {id: 'b'}, {id: 'a'}]}
  ].forEach(data => {
    it('sorts an string property', () => {
      const sorted = sut.toView(data.unsorted, 'id', data.dir);

      expect(sorted).toEqual(data.sorted);
    });
  });

  it('sorts a string case insensitive', () => {
    const unsorted = [{id: 'C'}, {id: 'a'}, {id: 'b'} ];
    const expectSort = [{id: 'a'}, {id: 'b'}, {id: 'C'} ];
    const actualSort = sut.toView(unsorted, 'id');

    expect(actualSort).toEqual(expectSort);
  });

  [ {
      dir: 'asc',
      unsorted: [{id: new Date(2012, 1, 6)}, {id: new Date(2011, 2, 5)}, {id: new Date(2012, 1, 1)} ],
      sorted: [{id: new Date(2011, 2, 5)}, {id: new Date(2012, 1, 1)}, {id: new Date(2012, 1, 6)}]
    },
    {
      dir: 'desc',
      unsorted: [{id: new Date(2012, 1, 6)}, {id: new Date(2011, 2, 5)}, {id: new Date(2012, 1, 1)} ],
      sorted: [{id: new Date(2012, 1, 6)}, {id: new Date(2012, 1, 1)}, {id: new Date(2011, 2, 5)}]
    }
  ].forEach(data => {
    it('sorts an date property', () => {
      const sorted = sut.toView(data.unsorted, 'id', data.dir);

      expect(sorted).toEqual(data.sorted);
    });
  });

  it('sorts a nested property', () => {
    const unsorted = [
      { a: 1, nested: { a: new Date(2015, 9, 2) } },
      { a: 2, nested: { a: new Date(2014, 9, 2) } }
    ];

    const sorted = sut.toView(unsorted, 'nested.a');

    expect(sorted).toEqual([
      { a: 2, nested: { a: new Date(2014, 9, 2) } },
      { a: 1, nested: { a: new Date(2015, 9, 2) } }
    ]);
  });
});


test/group.spec.js
import {GroupValueConverter} from '../src/group';

describe('the group by value converter', () => {
  let sut;
  let undefinedData = [null, undefined, ''];

  beforeEach(() => {
    sut = new GroupValueConverter();
  });

  undefinedData.forEach(val => {
    it('does nothing when no array is passed', () => {
      expect(() => sut.toView(val, 'g')).not.toThrow();
    });
  });

  undefinedData.forEach(val => {
    it('does nothing when no group by is passed', () => {
      expect(() => sut.toView([], val)).not.toThrow();
    });
  });

  // test items that would fail the if(!group) because i only want to keep
  // exclude undefined
  it('groups by property name', () => {
    const array = [
      { id: 1, group: 'a'},
      { id: 2, group: 'a'},
      { id: 3, group: null },
      { id: 4, group: false },
      { id: 5, group: '' },
      { id: 6 },
      { id: 7, group: undefined }
    ];

    const grouped = sut.toView(array, 'group');

    expect(grouped).toEqual([
      { group: 'a', values: [{ id: 1, group: 'a'}, { id: 2, group: 'a'}] },
      { group: '', values: [
        { id: 3, group: null },
        { id: 5,  group: '' },
        { id: 6 },
        { id: 7, group: undefined }
      ]},
      { group: 'false', values: [{ id: 4, group: false}] }
    ]);
  });

  it('groups by nested property name', () => {
    // add outer group so it targets correct prop
    const array = [
      { id: 1, group: 'x', nested: { group: 'a' } },
      { id: 2, group: 'y', nested: { group: 'a' } },
      { id: 3, group: 'z', nested: { group: 'b' } }
    ];

    const grouped = sut.toView(array, 'nested.group');

    expect(grouped).toEqual([{
      group: 'a',
      values: [
        { id: 1, group: 'x', nested: { group: 'a' }},
        { id: 2, group: 'y', nested: { group: 'a'}}
      ]}, {
        group: 'b',
        values: [{ id: 3, group: 'z', nested: { group: 'b' }}] }
    ]);
  });
});


config.js
System.config({
  defaultJSExtensions: true,
  transpiler: false,
  paths: {
    "github:*": "jspm_packages/github/*",
    "npm:*": "jspm_packages/npm/*"
  },

  map: {
    "aurelia-pal": "npm:aurelia-pal@1.3.0",
    "aurelia-polyfills": "npm:aurelia-polyfills@1.1.1",
    "lodash.topath": "npm:lodash.topath@4.5.2",
    "moment": "npm:moment@2.17.1",
    "github:jspm/nodelibs-assert@0.1.0": {
      "assert": "npm:assert@1.4.1"
    },
    "github:jspm/nodelibs-buffer@0.1.1": {
      "buffer": "npm:buffer@5.1.0"
    },
    "github:jspm/nodelibs-process@0.1.2": {
      "process": "npm:process@0.11.10"
    },
    "github:jspm/nodelibs-util@0.1.0": {
      "util": "npm:util@0.10.3"
    },
    "github:jspm/nodelibs-vm@0.1.0": {
      "vm-browserify": "npm:vm-browserify@0.0.4"
    },
    "npm:assert@1.4.1": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "buffer": "github:jspm/nodelibs-buffer@0.1.1",
      "process": "github:jspm/nodelibs-process@0.1.2",
      "util": "npm:util@0.10.3"
    },
    "npm:aurelia-polyfills@1.1.1": {
      "aurelia-pal": "npm:aurelia-pal@1.3.0"
    },
    "npm:buffer@5.1.0": {
      "base64-js": "npm:base64-js@1.3.0",
      "ieee754": "npm:ieee754@1.1.11"
    },
    "npm:inherits@2.0.1": {
      "util": "github:jspm/nodelibs-util@0.1.0"
    },
    "npm:lodash.topath@4.5.2": {
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:process@0.11.10": {
      "assert": "github:jspm/nodelibs-assert@0.1.0",
      "fs": "github:jspm/nodelibs-fs@0.1.2",
      "vm": "github:jspm/nodelibs-vm@0.1.0"
    },
    "npm:util@0.10.3": {
      "inherits": "npm:inherits@2.0.1",
      "process": "github:jspm/nodelibs-process@0.1.2"
    },
    "npm:vm-browserify@0.0.4": {
      "indexof": "npm:indexof@0.0.1"
    }
  }
});


src/sort.js
import { getObjectValue } from './utils';
import {
  dateComparison,
  isDate,
  isString
} from './utils.js';

function _isDesc(direction) {
  return direction.toLowerCase() === 'desc';
}

export class SortValueConverter {
  /**
   * @desc Sorts the array
   * @param {Array} array the array to sort
   * @param {String} [propertyName] the property name to group by
   * @param {String} [direction=asc] the direction to sort
   */
  toView(array, propertyName, direction) {
    if (!array) return;

    const isDesc = _isDesc(direction || 'asc');
    const factor = isDesc ? -1 : 1;

    return array.slice(0).sort((a, b) => {
      let aProp = getObjectValue(a, propertyName);
      let bProp = getObjectValue(b, propertyName);

      if (isDate(aProp) && isDate(bProp)) {
        return dateComparison(aProp, bProp) * factor;
      }

      if (isString(aProp)) aProp = aProp.toLowerCase();
      if (isString(bProp)) bProp = bProp.toLowerCase();

      return (aProp > bProp ? 1 : -1) * factor;
    });
  }
}



src/group.js
import { getObjectValue } from './utils';

export class GroupValueConverter {
  /**
   * @desc Groups the array by the property name
   * @param {Array} array the array to group
   * @param {String} propertyName the property name to group by
   */
  toView(array, propertyName) {
    if (!array || !propertyName) return array;

    const groups = { };

    // create the groups
    array.forEach(obj => {
      let group = getObjectValue(obj, propertyName);

      if (isMissing(group)) group = '';

      groups[group] = groups[group] || [];
      groups[group].push(obj);
    });

    // map out to a new object
    return Object.keys(groups).map(group => {
      return {
        group: group,
        values: groups[group]
      };
    });
  }
}

function isMissing(val) {
  return val === null || typeof val === 'undefined';
}


src/utils.js
import moment from 'moment';
import topath from 'lodash.topath';

/**
 * Compares two dates (Date objects or strings).
 * @param {Date} a the first date
 * @param {Date} b the second date
 * @return {Number} number representing the diff from a to b
 */
export function dateComparison(a, b) {
  if (a === null) a = new Date(1900, 0, 1);
  if (b === null) b = new Date(1900, 0, 1);
  return moment(a).diff(moment(b));
}

export function isDate(val) {
  return moment(val, moment.ISO_8601, true).isValid();
}

export function isString(val) {
  return typeof val === 'string';
}

export function getObjectValue(obj, propertyPath) {
  if (!obj) {
    throw new Error('Must provide an object to get the value');
  }
  if (!propertyPath) {
    return obj;
  }

  const paths = topath(propertyPath);
  const length = paths.length;

  let result = obj;
  for (let i = 0; i < length && result; ++i) {
    result = result[paths[i]];
  }

  return result;
}


karma.conf.js
// Karma configuration
// Generated on Fri Dec 05 2014 16:49:29 GMT-0500 (EST)

module.exports = function(config) {
  config.set({

    // base path that will be used to resolve all patterns (eg. files, exclude)
    basePath: '',

    // frameworks to use
    // available frameworks: https://npmjs.org/browse/keyword/karma-adapter
    frameworks: ['jspm', 'jasmine'],

    jspm: {
      // Edit this to your needs
      loadFiles: ['test/setup.js', 'test/**/*.js'],
      serveFiles: ['src/**/*.js'],
      paths: {
        '*': '*',
        'github:*': 'jspm_packages/github/*',
        'npm:*': 'jspm_packages/npm/*'
      }
    },

    // list of files / patterns to load in the browser
    files: [],

    // list of files to exclude
    exclude: [],


    // preprocess matching files before serving them to the browser
    // available preprocessors: https://npmjs.org/browse/keyword/karma-preprocessor
    preprocessors: {
      'test/**/*.js': ['babel'],
      'src/**/*.js': ['babel']
    },
    'babelPreprocessor': {
      options: {
        sourceMap: 'inline',
        presets: [ 'es2015-loose', 'stage-1'],
        plugins: [
          'syntax-flow',
          'transform-decorators-legacy',
          'transform-flow-strip-types'
        ]
      }
    },

    // test results reporter to use
    // possible values: 'dots', 'progress'
    // available reporters: https://npmjs.org/browse/keyword/karma-reporter
    reporters: ['progress'],

    // web server port
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    autoWatch: true,

    // start these browsers
    // available browser launchers: https://npmjs.org/browse/keyword/karma-launcher
    browsers: ['Chrome'],

    // Continuous Integration mode
    // if true, Karma captures browsers, runs the tests and exits
    singleRun: false
  });
};
